File: leetcode_solutions/0904-fruit-into-baskets/0904-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        mapping = {}
        left = 0
        right = 0
        max_fruit = float('-inf')
        while right < len(fruits):
            f = fruits[right]
            mapping[f] = mapping.get(f, 0) + 1
            if len(mapping) > 2:
                if mapping[fruits[left]] == 1:
                    del mapping[fruits[left]]
                else:
                    mapping[fruits[left]] -= 1
                left += 1
            
            max_fruit = max(max_fruit, right - left + 1)
            right += 1
        return max_fruit
    
    def totalFruitForLoop(self, fruits: List[int]) -> int:
        mapping = {}
        l = 0
        res = float('-inf')
        for r, f in enumerate(fruits):
            mapping[f] = mapping.get(f, 0) + 1
            if len(mapping) > 2:
                if mapping[fruits[l]] == 1: 
                    del mapping[fruits[l]]
                else: 
                    mapping[fruits[l]] -= 1
                l += 1
            res = max(res, r - l + 1)
        return res
        

File: leetcode_solutions/0904-fruit-into-baskets/0904-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        Map<Integer, Integer> map = new HashMap<>();
        int left = 0;
        int maxFruit = Integer.MIN_VALUE;
        int totalFruits = 0;
        for(int right = 0; right < fruits.length; right++){
            int fRight = fruits[right];
            map.put(fRight, map.getOrDefault(fRight, 0) + 1);
            totalFruits++;
            // Validate the widnow
            if(map.size() > 2){
                int fLeft = fruits[left];
                // Decrement left fruit from the hashmap
                map.put(fLeft, map.get(fLeft) - 1);
                // Decrement total fruits
                totalFruits--;
                
                // If the count of map happens to be zero
                if(map.get(fLeft) == 0){
                    // Pop the left key from the hashmap
                    map.remove(fLeft);
                }
                left++;
            }
            
            maxFruit = Math.max(maxFruit, totalFruits);
        }
        return maxFruit;
    }
}